napari/_qt/menus/plugins_menu.py

```diff
@@ -59,7 +59,15 @@ def _build(self, event=None):
                     full_name = menu_item_template.format(*key)
                     action = QAction(full_name, parent=self)
 
-                def _add_widget(*args, key=key, hook_type=hook_type):
+                def _add_toggle_widget(*args, key=key, hook_type=hook_type):
+                    full_name = menu_item_template.format(*key)
+                    if full_name in self._win._dock_widgets.keys():
+                        dock_widget = self._win._dock_widgets[full_name]
+                        if dock_widget.isVisible():
+                            dock_widget.hide()
+                        else:
+                            dock_widget.show()
+                        return
```

Comment:
Do you think it would be ok to make that a method on PluginsMenu ? It seem to be duplicated with `_add_registered_widget`. 

In general this whole block is a duplicate of `_add_registered_widget`, with the exception of `if event.value == plugin_name:`, so you can likely just call `_add_registered_widget` and modify it with

```
def _add_registered_widget(self, event=None, *, call_all=False):
...
    if event.value == plugin_name or call_call:
```

Maybe ? 

That should cut about 50 lines. 

Or you can make the whole inner logic of the for loop a utility function ? Up to you. 
I'm also ok to leave it as is.

Reply:
thanks @Carreau.  You are right. This whole part can be cleaned up a bit. I was having some trouble with it before, but I will revisit it and try to make this better.  thanks for the suggestions!